wcomponents-theme/src/main/js/wc/dom/getLabelsForElement.js

```diff
@@ -78,35 +100,37 @@ define(["wc/dom/tag",
 		 *  is returned.
 		 */
 		function getLabels(element, includeReadOnly) {
-			var result = [],
+			var ariaLabelledElements = [],
+				nativeLabeledElements = [],
 				label,
 				tagName;
+			if (element) {
+				ariaLabelledElements = getAriaLabelledElements(element);
+
+				if (wrappedInput.isOneOfMe(element, includeReadOnly)) {
+					nativeLabeledElements = getLabelsForWrapper(element, includeReadOnly);
+				} else {
+					FIELDSET = FIELDSET || new Widget(tag.FIELDSET);
+					if (FIELDSET.isOneOfMe(element)) {
+						LEGEND = LEGEND || new Widget("legend");
+						if ((label = LEGEND.findDescendant(element, true))) {
+							nativeLabeledElements = [label];
+						}
+					}
 
-			if (wrappedInput.isOneOfMe(element, includeReadOnly)) {
-				return getLabelsForWrapper(element, includeReadOnly);
-			}
+					nativeLabeledElements = doLabelQuery(element, nativeLabeledElements, includeReadOnly);
 
-			FIELDSET = FIELDSET || new Widget(tag.FIELDSET);
-			if (FIELDSET.isOneOfMe(element)) {
-				LEGEND = LEGEND || new Widget("legend");
-				if ((label = LEGEND.findDescendant(element, true))) {
-					result = [label];
+					if (!(nativeLabeledElements && nativeLabeledElements.length)) {
+						// try getting an ancestor label element ONLY if element is input, textarea, select or progress.
+						tagName = element.tagName;
+						if (LABELABLE.indexOf(tagName) > -1) {
```

Comment:
Do you really need the tag name var?

Reply:
Perhaps not required, it is existing code I refactored. Anyway agree, I have removed it now.